File: src/protection.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "protection.h"
#include "utils.h"
#include "encryption/sha256.h"
#include "encryption/aes.h"

int create_passphrase_digest(const unsigned char* passphrase, unsigned char* digest)
{
	int was_successful = 0;
	
	do{
		if(!passphrase)
			break;
			
		SHA256_CTX ctx;
		size_t passphrase_len = strlen((char*)passphrase);
		
		sha256_init(&ctx);
		sha256_update(&ctx, passphrase, passphrase_len);
		sha256_final(&ctx, digest);
		
		was_successful = 1;
	}while(0);

	if(!was_successful)
	{
	
	}

	return was_successful;
}


int encrypt_memory(const unsigned char* buffer,
				   size_t buffer_len,
				   const unsigned char* key,
				   unsigned char** encrypted_buffer,
				   size_t* encrypted_buffer_len)
{
	int was_successful = 0;
	unsigned int key_schedule[60] = {0};
	
	do{
		if(!buffer || !key || 0 >= buffer_len)
			break;
			
		*encrypted_buffer_len = buffer_len % 16;
		*encrypted_buffer_len = 16 - *encrypted_buffer_len;
		*encrypted_buffer_len += buffer_len;
		
		*encrypted_buffer = (unsigned char*) malloc(sizeof(unsigned char) * (*encrypted_buffer_len));
		if(!(*encrypted_buffer))
			break;
			
		memset(*encrypted_buffer, 0, *encrypted_buffer_len);
		
		const unsigned char* ptr_buffer = buffer;
		unsigned char* ptr_encrypted_buffer = *encrypted_buffer;
			
		aes_key_setup((unsigned char*)key, key_schedule, 256);
		
		int index = 0;
		int can_break = 0;
		do
		{
			unsigned char input[16];
			unsigned char cipher[16];
			
			memset(input, 0, 16);
			memset(cipher, 0, 16);
			
			if(buffer_len <= 0)
				break;
			
			if(buffer_len < 16)
			{
				memcpy(input, ptr_buffer + index, buffer_len);
				can_break = 1;
			}
			else
			{
				memcpy(input, ptr_buffer + index, 16);
				buffer_len -= 16;
				index += 16;
			}
			
			aes_encrypt(input, cipher, key_schedule, 256);	
			memcpy(ptr_encrypted_buffer, cipher, 16);

			ptr_encrypted_buffer += 16;
		}while(!can_break);
		
		was_successful = 1;
	}while(0);

	if(!was_successful)
	{
	}

	return was_successful;
}

int decrypt_memory(const unsigned char* buffer,
				   size_t buffer_len,
				   const unsigned char* key,
				   unsigned char** decrypted_buffer,
				   size_t* decrypted_buffer_len)
{
	int was_successful = 0;
	unsigned int key_schedule[60] = {0};
	
	do{
		if(!buffer || !key || 0 >= buffer_len)
			break;
			
		*decrypted_buffer_len = buffer_len % 16;
		*decrypted_buffer_len = 16 - *decrypted_buffer_len;
		*decrypted_buffer_len += buffer_len;
			
		*decrypted_buffer = (unsigned char*) malloc(sizeof(unsigned char) * (*decrypted_buffer_len));
		if(!(*decrypted_buffer))
			break;
			
		memset(*decrypted_buffer, 0, *decrypted_buffer_len);
		
		const unsigned char* ptr_buffer = buffer;
		unsigned char* ptr_decrypted_buffer = *decrypted_buffer;
			
		aes_key_setup(key, key_schedule, 256);
		
		int index = 0;
		int can_break = 0;
		do
		{
			unsigned char output[16];
			unsigned char cipher[16];
			
			memset(output, 0, 16);
			memset(cipher, 0, 16);
			
			if(buffer_len <= 0)
				break;
			
			if(buffer_len < 16)
			{
				memcpy(cipher, ptr_buffer + index, buffer_len);
				can_break = 1;
			}
			else
			{
				memcpy(cipher, ptr_buffer + index, 16);
				buffer_len -= 16;
				index += 16;
			}
						
			aes_decrypt(cipher, output, key_schedule, 256);
			memcpy(ptr_decrypted_buffer, output, 16);

			ptr_decrypted_buffer += 16;
		}while(!can_break);

		was_successful = 1;
	}while(0);
	

	if(!was_successful)
	{
	}

	return was_successful;
}

File: src/protection.nim
{.compile: "c/aes.c".}
{.compile: "c/sha256.c".}
{.compile: "c/utils.c".}

type TDigest* = array[0..31, uint8]

type TProtectedBuffer = object
  buffer*: cstring
  len*: cint

proc c_createPassphraseDigest(passphrase: cstring, 
                              digest: var TDigest): 
  cint {.importc: "create_passphrase_digest".}
proc c_encryptMemory( buffer: cstring, 
                      bufferLen: cint, 
                      key: TDigest, 
                      encryptedBuffer: var cstring, 
                      encryptedBufferLen: var cint ): 
  cint {.importc: "encrypt_memory".}
proc c_decryptMemory( buffer: cstring, 
                      bufferLen: cint, 
                      key: TDigest, 
                      decryptedBuffer: var cstring, 
                      decryptedBufferLen: var cint ): 
  cint {.importc: "decrypt_memory".}

proc c_saveEncryptedMemory( path: cstring,
                            buffer: cstring,
                            bufferLen: cint):
  cint {.importc: "save_encrypted_memory".}

proc c_loadEncryptedFile( path: cstring,
                          key: TDigest,
                          decryptedBuffer: var cstring,
                          decryptedBufferLen: var cint):
  cint {.importc: "load_encrypted_file".}

proc c_freeAllocatedBuffer(buffer: cstring): 
  cint {.importc: "free_allocated_mem".} # not sure if we need this one

## convert TDigest to string
proc toString*(digest: TDigest): string = 
  result = ""
  for index in 0..31:
    result &= $digest[index]

## convert TDigest to string
proc `$`*(digest: TDigest): string = 
  result = toString(digest)

## produce digest out of passphrase
proc createPassphraseDigest*(passphrase: cstring): TDigest = 
  discard c_createPassphraseDigest(passphrase, result)

## generate protected buffer object out of raw string
proc createProtectedBuffer*(buffer: cstring, len: cint): ref TProtectedBuffer = 
  new(result)
  result.buffer = buffer
  result.len = len

## encrypt buffer
proc encryptBuffer*(buffer: cstring, key: TDigest): ref TProtectedBuffer = 
  var bufferLen = cast[cint](buffer.len)

  new(result)
  result.buffer = nil
  result.len = 0
    
  if 0 == c_encryptMemory(buffer, bufferLen, key, result.buffer, result.len):
    echo("Error: encryptBuffer-1")
    return nil
  if 0 == result.len:
    echo("Error: encryptBuffer-1")
    return nil

## decrypt buffer
proc decryptBuffer*(protectedBuffer: ref TProtectedBuffer, key: TDigest): string = 
  var decBuffer: cstring = nil
  var decBufferLen: cint = 0

  if 0 == c_decryptMemory(protectedBuffer.buffer, protectedBuffer.len, key, decBuffer, decBufferLen):
    echo("Error: decryptBuffer-1")
    return nil
  if 0 == decBufferLen:
    echo("Error: decryptBuffer-2")
    return nil

  result = $decBuffer
  discard c_freeAllocatedBuffer(decBuffer)

## save encrypted buffer to file
proc saveEncryptedBuffer*(path: cstring, protectedBuffer: ref TProtectedBuffer): bool =
  if 0 == c_saveEncryptedMemory(path, protectedBuffer.buffer, protectedBuffer.len):
    echo("Error: failed to save encrypted buffer")
    return false

  return true

## loads an encrypted buffer and returns the decrypted value  
proc decryptFile*( path: cstring,
                  key: TDigest): string = 
  var decBuffer: cstring = nil
  var decBufferLen: cint = 0

  if 0 == c_loadEncryptedFile(path, key, decBuffer, decBufferLen):
    echo("Error: failed to load encrypted file");
    return nil

  result = $decBuffer
  discard c_freeAllocatedBuffer(decBuffer)

when isMainModule:
  var buffer = """{ "title": "github3", "username": "sepisoad", "password": "1234", "email": "UNDEFINED", "description": "UNDEFINED", "date": "2015-03-01", }"""
  var key = createPassphraseDigest("hello")
  var encBuffer = encryptBuffer(buffer, key)

  if false == saveEncryptedBuffer("DB", encBuffer):
    echo ("OH SHIT")

  var decBuffer = decryptFile("DB", key)
  if nil == decBuffer:
    echo ("OH FUCK")

  echo(decBuffer)
